lightning/src/ln/msgs.rs

```diff
@@ -1934,20 +1937,39 @@ impl Writeable for OnionHopData {
 					(6, short_channel_id, required)
 				});
 			},
-			OnionHopDataFormat::FinalNode { ref payment_data, ref payment_metadata, ref keysend_preimage } => {
+			OnionHopDataFormat::FinalNode { ref payment_data, ref payment_metadata, ref keysend_preimage, ref custom_tlvs } => {
+				let mut custom_tlv_stream = custom_tlvs.clone().unwrap_or(Vec::new());
+				encode_tlv_stream!(&mut custom_tlv_stream, {
+					(5482373484, keysend_preimage, option)
+				});
+				let sorted_custom_tlvs = sorted_tlv_stream(&custom_tlv_stream)?;
 				_encode_varint_length_prefixed_tlv!(w, {
 					(2, HighZeroBytesDroppedBigSize(self.amt_to_forward), required),
 					(4, HighZeroBytesDroppedBigSize(self.outgoing_cltv_value), required),
 					(8, payment_data, option),
-					(16, payment_metadata.as_ref().map(|m| WithoutLength(m)), option),
-					(5482373484, keysend_preimage, option)
-				});
+					(16, payment_metadata.as_ref().map(|m| WithoutLength(m)), option)
+				}, &sorted_custom_tlvs);
 			},
 		}
 		Ok(())
 	}
 }
 
+fn sorted_tlv_stream(bytes: &[u8]) -> Result<Vec<u8>, io::Error> {
+	let mut tlv_stream = TlvStream::try_new(bytes).map_err(|err| {
+		match err {
+			DecodeError::Io(kind) => io::Error::new(kind, kind.to_string()),
+			_ => io::Error::new(io::ErrorKind::Other, err.to_string()),
+		}
+	})?.collect::<Vec<TlvRecord>>();
```

Comment:
Since writes are usually `expect/unwrap`'d, if custom TLVs are invalid here it'll panic while serializing `OnionHopData`. The user shouldn't be able to provide an invalid TLV stream since `RecipientOnionFields::with_custom_tlvs` should prevent that, but should we allow this? Thinking maybe should I add a check somewhere in the payment flow to return an error before this, or maybe on failure, omit the custom tlvs and go through with the payment?

Reply:
Became unnecessary when changing custom tlvs from a `Vec<u8>` to `Vec<(u64, Vec<u8>)>`